### PR TITLE
Initial code for kubernetes integration

### LINKED ISSUE #61: Read OAuth Tokens from file (Kubernetes integration)
Support reading OAuth tokens from file `${CREDENTIALS_DIR}/${token_name}-token-secret`, e.g. `/meta/credentials/full-access-token-secret`.

This is needed to support the OAuth/IAM integration of Kubernetes: https://github.com/zalando-incubator/kubernetes-on-aws/pull/154

### PR DESCRIPTION
**Pull request by jbellmann**
fixes #61

### CODE REVIEW

File: src/main/java/org/zalando/stups/tokens/k8s/KubernetesAccessTokenRefresher.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    //@formatter:off
+    @Override
+    public void start() {
+        initializeFixedTokensFromEnvironment();
+        LOG.info("Starting to refresh tokens regularly...");
+
+        FilesystemReader reader = new FilesystemReader(accessTokens);
+
+        if(configuration.getKubernetesConfiguration().isValidateTokensOnStartup()){
```
Comment (hjacobs) [inline src/main/java/org/zalando/stups/tokens/k8s/KubernetesAccessTokenRefresher.java]: nitpick: missing space after "if"

**Comment by hjacobs** [comment]
:+1: for dropping Java 7.

Maybe @harti2006 could have a look..

File: src/main/java/org/zalando/stups/tokens/k8s/FilesystemReader.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        }
+    }
+
+    //@formatter:off
+    protected void readFromFilesystem() {
+        LOG.debug("read from filesystem ...");
+        final File[] tokenSecretFiles = getCredentialsDir().listFiles(endsWithSuffixFilenameFilter);
+        asList(tokenSecretFiles)
+                .stream()
+                .map(f -> buildAccessTokenDto(f))
```
Comment (harti2006) [inline src/main/java/org/zalando/stups/tokens/k8s/FilesystemReader.java]: Java 8 hint: can be replaced by method reference:

`.map(this::buildAccessTokenDto)`

File: src/main/java/org/zalando/stups/tokens/k8s/FilesystemReader.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                });
+    }
+    //@formatter:on
+
+    protected AccessTokenDto buildAccessTokenDto(File tokenSecretFile) {
+        final String name = tokenSecretFile.getName().replace(TOKEN_SECRET_SUFFIX, "");
+
+        try {
+            String secret = FileUtils.readContent(tokenSecretFile.getAbsolutePath());
+            String type = FileUtils.readContent(tokenTypeFilePath(tokenSecretFile.getParentFile(), name));
```
Comment (harti2006) [inline src/main/java/org/zalando/stups/tokens/k8s/FilesystemReader.java]: There is java.nio.file.Path interface, that offers a `resolveSibling` method. Might be useful here. See
https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#resolveSibling-java.lang.String-

File: src/main/java/org/zalando/stups/tokens/k8s/KubernetesAccessTokenRefresher.java (hunk 1 above)
Comment (harti2006) [inline src/main/java/org/zalando/stups/tokens/k8s/KubernetesAccessTokenRefresher.java]: That's the burden of `//@formatter:off` 😛

File: src/main/java/org/zalando/stups/tokens/k8s/FilesystemReader.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    //@formatter:off
+    protected void readFromFilesystem() {
+        LOG.debug("read from filesystem ...");
+        final File[] tokenSecretFiles = getCredentialsDir().listFiles(endsWithSuffixFilenameFilter);
+        asList(tokenSecretFiles)
+                .stream()
+                .map(f -> buildAccessTokenDto(f))
+                .filter(Objects::nonNull)
+                .collect(toList())
```
Comment (harti2006) [inline src/main/java/org/zalando/stups/tokens/k8s/FilesystemReader.java]: This is obsolete, as you can invoke `.forEach` on the stream directly

File: src/main/java/org/zalando/stups/tokens/k8s/FilesystemReader.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        } catch (Exception e) {
+            LOG.error(e.getMessage(), e);
+        }
+    }
+
+    //@formatter:off
+    protected void readFromFilesystem() {
+        LOG.debug("read from filesystem ...");
+        final File[] tokenSecretFiles = getCredentialsDir().listFiles(endsWithSuffixFilenameFilter);
+        asList(tokenSecretFiles)
```
Comment (harti2006) [inline src/main/java/org/zalando/stups/tokens/k8s/FilesystemReader.java]: asList can be omitted by using `Stream.of(T...)`

File: src/main/java/org/zalando/stups/tokens/k8s/KubernetesAccessTokenRefresher.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import static java.util.stream.Collectors.toList;
+
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.zalando.stups.tokens.AbstractAccessTokenRefresher;
+import org.zalando.stups.tokens.TokenRefresherConfiguration;
+
+public class KubernetesAccessTokenRefresher extends AbstractAccessTokenRefresher {
```
Comment (harti2006) [inline src/main/java/org/zalando/stups/tokens/k8s/KubernetesAccessTokenRefresher.java]: I wouldn't necessarily mention "Kubernetes" at all, as there is nothing K8s-special in the approach. It's just about reading tokens from files.

Comment (jbellmann) [inline src/main/java/org/zalando/stups/tokens/k8s/KubernetesAccessTokenRefresher.java]: Ok, come with a better one. :-)

Comment (harti2006) [inline src/main/java/org/zalando/stups/tokens/k8s/KubernetesAccessTokenRefresher.java]: Naming is the hardest IT problem 😆 
What about "FilesAccessTokenRefresher", "SecretFilesAccessTokenRefresher", "FilesystemAccessTokenRefresher" or something similar.

**Comment by harti2006** [comment]
Would be great to have some sentences about the new feature in the readme and in the Java docs

**Comment by hjacobs** [comment]
@zmeda maybe you want to have a look too? (this PR is about Kubernetes integration)

**Comment by zmeda** [comment]
@hjacobs I will have a look and test it against current Kubernetes setup for our service.

File: src/test/java/org/zalando/stups/tokens/fs/FilesystemSecretRefresherTest.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.zalando.stups.tokens.AccessTokens;
+import org.zalando.stups.tokens.Tokens;
+
+public class FilesystemSecretRefresherTest {
+
+    @Before
+    public void setup() {
+        System.getProperties().put("CREDENTIALS_DIR", "k8s");
```
Comment (harti2006) [inline src/test/java/org/zalando/stups/tokens/fs/FilesystemSecretRefresherTest.java]: "k8s", "kube" and "kubernetes" still appear quite often in the code ;-)

**Comment by zmeda** [comment]
The changes were tested in Kubernetes and it seems like it works fine (with the `gerry` approach as well as with new credentials approach)

**Comment by hjacobs** [comment]
:+1:

**Comment by hjacobs** [comment]
@jbellmann could you merge & release?

**Comment by zmeda** [comment]
👍

**Comment by hjacobs** [comment]
What's the status here? Is it released, which version? Can we have release notes here: https://github.com/zalando/tokens/releases ?

**Comment by duergner** [comment]
As I already did some JavaDoc here I could try and add some more on the new classes as well. I'd just vote to do it with an additional PR not on this one as it's already quite big anyway and @hjacobs wants that one to be merged and released sooner than later I assume.

**Comment by hjacobs** [comment]
@duergner sounds great!

**Comment by hjacobs** [comment]
:+1:

**Comment by harti2006** [comment]
:+1:

**Pull request merged**